drivers/hud_apr/app/models/hud_apr/generators/shared/fy2024/base.rb

```diff
@@ -0,0 +1,1112 @@
+    # client in the table above.
+    # 5. If, for a given client, none of the records found belong to the same [project id] as the CE assessment from step 1, use the latest of those to report the client in the table above.
+    # 6. The intention of the criteria is to locate the most recent logically relevant record pertaining to the CE assessment record reported in Q9a and Q9b by giving preference to data entered by the same project.
+    private def latest_ce_event(she_enrollment, hoh_enrollment, ce_latest_assessment)
+      # need first assessment after report end if it occurred within 90 days of report end
+      # exclude events after that assessment if it exists
+      enrollment = if she_enrollment.client.source_events.present?
+        she_enrollment
+      else
+        hoh_enrollment
+      end
+      return unless enrollment.present?
+
+      potential_events = enrollment.client.source_events.select do |e|
+        next_assessment = first_ce_assessment_within_90_days_after_report_range(she_enrollment)
+        if ce_latest_assessment
+          start_date_check = ce_latest_assessment.AssessmentDate
+          end_date_check = [@report.end_date + 90.days, next_assessment&.AssessmentDate].compact.min
+        else
+          start_date_check = @report.start_date
+          end_date_check = @report.end_date
+        end
+        e.EventDate.present? && e.EventDate.between?(
+          start_date_check,
+          end_date_check,
+        )
+      end
+      events_from_project = potential_events.select do |e|
+        e.enrollment.project.id == she_enrollment.project.id
+      end
+      return events_from_project.max_by(&:EventDate) if events_from_project.present?
+
+      potential_events.max_by(&:EventDate)
+    end
+
+    private def apr_age_ranges
+      @apr_age_ranges ||= age_ranges.select { |k, _| age_range_keys.include?(k) }
+    end
+
+    private def age_range_keys
+      [
+        'Under 5',
+        '5-12',
+        '13-17',
+        '18-24',
+        '25-34',
+        '35-44',
+        '45-54',
+        '55-64',
+        '65+',
+        label_for(:dkptr),
+        'Data Not Collected',
+        'Total',
+      ].freeze
+    end
+
+    private def age_ranges
+      {
+        'Under 5' => a_t[:age].between(0..4),
+        '5-12' => a_t[:age].between(5..12),
+        '13-17' => a_t[:age].between(13..17),
+        '18-24' => a_t[:age].between(18..24),
+        '25-34' => a_t[:age].between(25..34),
+        '35-44' => a_t[:age].between(35..44),
+        '45-54' => a_t[:age].between(45..54),
+        '55-64' => a_t[:age].between(55..64),
+        '65+' => a_t[:age].gteq(65),
+        label_for(:dkptr) => a_t[:dob_quality].in([8, 9]).and(a_t[:dob].eq(nil)),
+        'Data Not Collected' => a_t[:dob_quality].eq(99).and(a_t[:dob].eq(nil)),
+        'Total' => Arel.sql('1=1'), # include everyone
+      }
+    end
```

Comment:
Should this live in `app/models/concerns/hud_reports/ages.rb` ?  Does it not already live there?